### DISCUSSION TITLE
Execute some hooks when images are added/removed

### DISCUSSION THREAD

**Discussion by arouiadib** [PrestaShop/PrestaShop#33411] [Ideas]
Images is an important topic when merchant has a lot of products, especially if he uses high size images and large image types. In this case, storing them in local drive is not cost effective and not reliabe.

Using cloud storage services (like Azure or S3 to name of few) is a good solution.They also improve performance by offloading web server from image requests.

Also migrating to another host is more easy, because the files backup is smaller. One approach is to keep only `tmp` folder inside `img` folder (just for backoffice images)  and upload all the rest to the cloud.

The problem is for adding/ removal of images. There should be a syncing mechanism between local drive and cloud.

It is possible to upload product image to the cloud just after the upload to local drive using the hook `actionWatermark`.

But for deletion, I don’t find a hook that executes once a product image is deleted.

For the other images (manufacturers, suppliers, shipping... ) I dont find any hooks executed on add/removal of images.

I understand that the most important images in size are product images (and maybe manufacturers). But when we activate media servers in `Advanced >performance`, all images in front office are from cloud, which emphasizes the need of a syncing mechanism also for categories, suppliers..etc.

**Suggestions:**
Add hook actionProductImageDeleted
Add hook actionManufacturerImageAdded
Add hook actionManufacturerImageDeleted
Add other similar hooks for categories….

I would like to know pros/cons of adding such hooks.
Community feedback is much appreciated.

**Comment by sarahdib** [comment]
@PrestaShop/prestashop-core-developers could you check this improvement ?

Thank you :)

**Comment by matks** [comment]
Hi @arouiadib we have entered the month of April which is the month where PrestaShop 1.7.8 should be feature frozen. The very last Pull Requests are being submitted and validated so we can complete the 1.7.8 milestone. So answers might come late as maintainers are very busy at this moment! 😅 

I think the idea is good nonetheless

**Comment by arouiadib** [comment]
I totally understand @matks, and would like to thank you and all the maintainers for the huge effort.

This matter itself is not urgent so we will be discussing it whenever there is appropriate moment.

I started by a PR regarding the first hook (and most important one IMO). I will be writing the other PRs gradually when community confirm to me each time (because not all hooks have same importance).

Thanks

**Comment by kpodemski** [comment]
I suggest to check dynamic hooks available in ObjectModel::delete(), using them should be enough.

**Comment by arouiadib** [comment]
Thank you @kpodemski for information about these hooks, new to me.

I think product images are the only images that are represented by an object model in PS (class `Image`). 
I don't think there is model object for manufacturer image. This image is named using the id of the manufacturer, and it is not persisted to database. This same fact applies to category, supplier...etc.

For product image, I agree that using the dynamic hook available in `ObjectModel::delete()` is a solution. One thing is that this hook happens before the real file deletion on hard disk :thinking: .

Executing a hook after all the deletion has happened (deletion from database and deletion of image from hard disk) sounds better to me. wdyt?

**Comment by cdcvince** [comment]
Actually I'm not sure that a hook will be enough, or these hooks need to be added in the Image Manager class and should have an impact on the code execution that follows.

The fact is that we don't want to add an action, but we want to replace it (upload / delete images in remote folder instead of in local folder).

But I agree that this feature is really important if we want that Prestashop fits for large shops.

**Comment by arouiadib** [comment]
@cdcvince thanks a lot, your comment is insightful and will be taken into account while looking for the better solution.

WIP, I will be back to feedback.
